server/conn.go

```diff
@@ -1480,7 +1480,8 @@ func (cc *clientConn) handleQuery(ctx context.Context, sql string) (err error) {
 		}
 		retryable, err = cc.handleStmt(ctx, stmt, parserWarns, i == len(stmts)-1)
 		if err != nil {
-			if cc.ctx.GetSessionVars().EnableTiFlashFallbackTiKV && errors.ErrorEqual(err, tikv.ErrTiFlashServerTimeout) && retryable {
+			_, allowTiFlashFallback := cc.ctx.GetSessionVars().AllowFallbackToTiKV[kv.TiFlash]
+			if allowTiFlashFallback && errors.ErrorEqual(err, tikv.ErrTiFlashServerTimeout) && retryable {
```

Comment:
1. could add some test that the query is not retryable but executed on TiFlash?
2. if the statement is retryable, but it's not executed on TiFlash, would it be retried as before?

Reply:
1. We have a test case where the query is not retryable but executed on TiFlash. https://github.com/pingcap/tidb/blob/master/server/conn_test.go#L775. I will add more.
2. The return value `retryable` is only used for TiFlash fallback. If the statement is retryable, but it's not executed on TiFlash, it would be retried as usual.

Reply:
so what's the difference between `allowTiFlashFallback` and `retryable`?

Reply:
`allowTiFlashFallback` is a switch to control whether TiFlash would fallback. When the switch is on, there are two cases which are hard to retry with TiKV. The first case is that TiDB has already sent packet to client, which is hard to revert. The second case is in cursor mode, which is also hard to revert the cursor. `retryable` will be false for the two cases.